File: typescript/src/com/murex/tbw/client/BooksOrder.ts
import { Order } from "./Order";
import { Book } from "../domain/book/Book";
import { Invoice } from "../purchase/Invoice";
import { Client } from "./Client";
import * as Immutable from "immutable";
import { PurchasedBook } from "../purchase/PurchasedBook";
import { MainRepository } from "../MainRepository";

export class BooksOrder implements Order {
  private readonly client: Client;
  private readonly booksInBasket = new Map<Book, number>();

  constructor(client: Client) {
    this.client = client;
  }

  addBook(book: Book, quantity: number): void {
    const alreadySold = this.booksInBasket.get(book) || 0;
    this.booksInBasket.set(book, alreadySold + quantity);
  }

  checkOut(): Invoice {
    const invoice = new Invoice(this.client.name, this.client.country);

    this.booksInBasket.forEach((quantity: number, book: Book) => {
      invoice.addPurchasedBook(new PurchasedBook(book, quantity));
    });
    MainRepository.configuredRepository().addInvoice(invoice);
    return invoice;
  }

  getQuantityOf(book: Book): number {
    return this.booksInBasket.get(book) || 0;
  }
}

File: scala/src/main/scala/com/murex/tbw/client/BooksOrder.scala
package com.murex.tbw.client

import com.murex.tbw.MainRepository
import com.murex.tbw.domain.book.Book
import com.murex.tbw.purchase.{Invoice, PurchasedBook}

class BooksOrder(private val client: Client) extends Order {
  private val booksInBasket: scala.collection.mutable.Map[Book, Int] =
    scala.collection.mutable.Map[Book, Int]()

  override def addBook(book: Book, quantity: Int): Unit = {
    booksInBasket(book) = quantity
  }

  override def checkOut(): Invoice = {
    val invoice = new Invoice(client.name, client.country)
    booksInBasket.foreach {
      case (book, quantity) => invoice.addBook(new PurchasedBook(book, quantity))
    }
    MainRepository.configuredRepository().addInvoice(invoice)
    invoice
  }

  override def getQuantityOf(book: Book): Int = booksInBasket.getOrElse(book, 0)
}
